EIPS/eip-6384.md

```diff
@@ -0,0 +1,155 @@
+---
+eip: 6384
+title: Humanly readable offline signatures
+description: A procedure for making a proposed to be signed buffer of typed structured data humanly readable.
+author: Tal Be'ery (@talbeerysec), RoiV (@DeVaz1)
+discussions-to: https://ethereum-magicians.org/t/eip-6384-readable-eip-712-signatures/12752
+status: Draft
+type: Standards Track
+category: Interface
+created: 2023-01-08
+requires: 712
+---
+
+## Abstract
+
+This is a standard for making a proposed to be signed typed structured data humanly readable. It includes
+
+- Specification for including a function in the verifying contract that returns a humanly readable description of the message to be signed,
+- reference implementation,
+- Security analysis of the proposed solution
+
+## Motivation
+
+The use case of Web3 off-chain signatures intended to be used within on-chain transaction is gaining traction and being used in multiple leading protocols (e.g. OpenSea) and standards [EIP-2612](./eip-2612.md), mainly as it offers a fee-less experience.
+Attackers are known to actively and successfully abuse such off-chain signatures, leveraging the fact that users are blindly signing off-chain messages, since they are not humanly readable.
+While [EIP-712](./eip-712.md) originally declared in its title that being ”humanly readable” is one of its goals, it did not live up to its promise eventually and EIP-712 messages are not understandable by an average user.
+
+The example below shows the user exprience of victims that browse an actual phishing website. The phishing website requests the victims to sign a message that will put their NFT token for sale on OpenSea platform for a virtually zero price.
+
+![](../assets/eip-6384/media/MiceyMask.gif)
+
+As this example shows, the user interface for some popular wallet implementations is not coveying the actual meaning of signing such transactions.
+
+In this proposal we offer a secure and scalable method to bring true human readability to EIP-712 messages by leveraging their binded smart contracts.
+As a result, once implemented this EIP wallets can upgrade their user experience from current state:
+
+![](../assets/eip-6384/media/MiceyMask-non-compliant.png)
+
+to a much clearer user experience:
+
+![](../assets/eip-6384/media/ZenGo-EIP-compliant.png)
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+EIP-712 already formally binds an off-chain signature to a contract, with the "verifyingContract" parameter. We suggest adding a “view” function ("stateMutability":"view") to such contracts, that returns a human readable description of the meaning of this specific off-chain buffer.
+
+```
+/**
+ * @dev Returns the expected result of the offchain message.
+*/
+
+     function evalEIP712Buffer(bytes32 domainHash, string memory primaryType, bytes memory signature)
```

Comment:
I would normally consider something called `signature` to be the `v`, `r`, and `s` values of an ECDSA signature. In this case, is `signature` the buffer of typed data? Probably should rename this parameter if that's the case.

Reply:
i've renamed the parameter name from signature to typedDataBuffer